### PR TITLE
[SPARK-20768][PYSPARK][ML] Expose numPartitions (expert) param of PySpark FPGrowth.

### PR DESCRIPTION
**Pull request by facaiy**
## What changes were proposed in this pull request?

Expose numPartitions (expert) param of PySpark FPGrowth.

## How was this patch tested?

+ [x] Pass all unit tests.

### CODE REVIEW

**Comment by yanboliang** [comment]
ok to test.

**Comment by facaiy** [comment]
Thanks, @yanboliang. 
Do you have any suggestion about testing the parameter?

**Comment by yanboliang** [comment]
Jenkins, test this please.

**Comment by yanboliang** [comment]
add to whitelist

**Comment by yanboliang** [comment]
ok to test

**Comment by yanboliang** [comment]
Jenkins, test this please.

**Comment by yanboliang** [comment]
@srowen @MLnick Could you help to add @facaiy to whitelist? It seems we can't trigger this job currently. Thanks.

**Comment by facaiy** [comment]
There seems something wrong with CI. I saw the same  non-response/delay of CI once again since last month.

**Comment by felixcheung** [comment]
Jenkins, ok to test

File: python/pyspark/ml/fpm.py [hunk 1]
```diff
@@ -49,6 +49,32 @@ def getMinSupport(self):
         return self.getOrDefault(self.minSupport)
 
 
+class HasNumPartitions(Params):
+    """
+    Mixin for param support.
```
Comment (yanboliang) [inline python/pyspark/ml/fpm.py]: ```Mixin for param numPartitions: Number of partitions (at least 1) used by parallel FP-growth.```

**Review by yanboliang — COMMENTED** [review]
One minor comment, otherwise, LGTM. Thanks.

File: python/pyspark/ml/fpm.py [hunk 2]
```diff
@@ -49,6 +49,32 @@ def getMinSupport(self):
...
+    """
+
+    numPartitions = Param(
+        Params._dummy(),
+        "numPartitions",
+        """Number of partitions (at least 1) used by parallel FP-growth.
```
Comment (yanboliang) [inline python/pyspark/ml/fpm.py]: Using ```"""``` to wrap doc here will get ```\n``` in generated Python API docs. You can update to use ```"``` referring to [discussion at here](https://github.com/apache/spark/pull/18089#discussion_r118280610).

File: python/pyspark/ml/fpm.py [hunk 3]
```diff
@@ -49,6 +49,32 @@ def getMinSupport(self):
...
+
+    def setNumPartitions(self, value):
+        """
+        Sets the value of :py:attr:`numPartitions`.
+        """
+        return self._set(numPartitions=value)
+
+    def getNumPartitions(self):
+        """
+        Gets the value of numPartitions or its default value.
```
Comment (felixcheung) [inline python/pyspark/ml/fpm.py]: `:py:attr:`

File: python/pyspark/ml/fpm.py (hunk 2 above)
Comment (felixcheung) [inline python/pyspark/ml/fpm.py]: does this need to be scrubbed ? I think we have `"""` everywhere

File: python/pyspark/ml/fpm.py (hunk 1 above)
Comment (facaiy) [inline python/pyspark/ml/fpm.py]: modified.

File: python/pyspark/ml/fpm.py (hunk 2 above)
Comment (facaiy) [inline python/pyspark/ml/fpm.py]: replaced.

File: python/pyspark/ml/fpm.py (hunk 3 above)
Comment (facaiy) [inline python/pyspark/ml/fpm.py]: added.

**Comment by facaiy** [comment]
Hi, I'm not familiar with pyspark. I just wonder whether is it needed to create a unit test for verification. If yes, how to check it? Thanks.

**Comment by yanboliang** [comment]
@facaiy Could you resolve merge conflicts? Then I can get this in. Thanks.

**Review by MLnick — APPROVED** [review]

**Comment by facaiy** [comment]
Resolved.

By the way,
Which one is preferable, rebase or merge?

**Comment by MLnick** [comment]
I personally prefer merging when the PR is still in progress - it preserves the commit history for reviewers.

**Comment by yanboliang** [comment]
Merged into master and branch-2.2. Thanks for all.

**Pull request closed without merging**